### ISSUE TITLE
Upstream HTTP request error

### ISSUE THREAD

**Issue by brandonkal** [openfaas-incubator/of-watchdog#50]
I am using OpenFaaS for a long running process. (Currently 3 minute jobs but plan to scale this up to 2 hour jobs).

I am using async-function to queue these jobs with a callback url.
I am also using the golang-http template.

```
2019/02/20 01:04:16 stderr: 2019/02/20 01:04:16 Starting job
2019/02/20 01:06:55 Upstream HTTP request error: Post http://127.0.0.1:8081/: EOF
```

It is not clear to me why this is set to 127.0.0.1:8081 as no docker containers are running on this port. I assume that this is the port the of-watchdog go process is running within the container, hence I am opening the issue here.

The callback url is receiving an empty POST with 503 as the function status code.
This does not occur when the handler returns a response earlier (e.g. if there is a 4xx error before processing).

## Expected Behaviour
Callback URL receives the data the handler returns.

## Current Behaviour
It appears the Upstream HTTP request causes an error, which then causes the POST to the callback url to be empty.

## Steps to Reproduce (for bugs)

1. Start with a golang-http template.
2. Long running process in Handler
3. Set environment: exec_timeout: 86400s in the stack.yml file (0s doesn't work)
4. Invoke function with a callback Url
5. tail the function logs and observe the POST error
6. Observe empty POST on callback server

## Your Environment

* Docker version `docker version` (e.g. Docker 17.0.05 ):
```
Client:
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.6
 Git commit:        6247962
 Built:             Sun Feb 10 04:13:47 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       6247962
  Built:            Sun Feb 10 03:42:13 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

* Are you using Docker Swarm or Kubernetes (FaaS-netes)?
Docker Swarm

* Operating System and version (e.g. Linux, Windows, MacOS):
Ubuntu 18.04

**Comment by burtonr** [comment]
I experienced this just today as well. This is an issue with Go and how the connections are handled. Sorry I can't provide a solution off-hand, but know you're not the only one! 

In my case, I am using the go-http template as the process function where this function calls out to 3 others to complete a workflow. I receive that error on the last of the 3 calls. The call succeeds, but the error is still present.

Here's what I was able to find on [StackOverflow](https://stackoverflow.com/questions/17714494/golang-http-request-results-in-eof-errors-when-making-multiple-requests-successi/23963271#23963271)
To summarize:
> The problem here occurs when you initiate another request before the readLoop actually detects the close, and the EOF that it reads get interpreted as an error for that new request rather than a close that occurred prior to the request.

**Comment by brandonkal** [comment]
@burtonr were you able to resolve this?

It looks like the request is occuring here rather than in my handler code:
https://github.com/openfaas-incubator/of-watchdog/blob/85505a7210cf413e455f8a03d74ba94d9a9fcd30/executor/http_runner.go#L119-L134

**Comment by burtonr** [comment]
I was not able to solve this. Since it's the last function call, and the call still succeeds, I moved on. Most of the team on this project are in the European time zones, so I was going to ask when they were available.

Based on that StackOverflow answer, I would say that you're probably right. Adding `request.Close()` there _may_ fix it. I'll give it a try and see what happens!

**Comment by brandonkal** [comment]
I have added `request.Close = true` below line 124 and that did not resolve the problem.

**Comment by brandonkal** [comment]
@burtonr After a bunch of debugging it appears the issue is occurring here: 

https://github.com/openfaas-incubator/golang-http-template/blob/4783533d117fcd65d16aa8557aff38e164a889e0/template/golang-http/main.go#L63-L64
```
ReadTimeout:  3 * time.Second,
WriteTimeout: 3 * time.Second,
```

That value is currently hard coded, but when I increase it I get this:
```
2019/02/20 05:20:52 POST / - 200 OK - ContentLength: 33
```
instead of this:
```
2019/02/20 05:15:59 Upstream HTTP request error: Post http://127.0.0.1:8081/: EOF
```
Perhaps this will help you.

I am not sure why that connection has to persist throughout the function execution. It would seem the of-watchdog process should just create a new connection when it needs to send data.

**Comment by burtonr** [comment]
Interesting. What did you set those values to? 

Those values should be configurable via environment variables `read_timeout` and `write_timeout` 

I'll wait until @alexellis comes back online to check with him. I'm sure he can provide better input.

I know this template is one of the more popular ones, and I haven't heard of this issue before. It's possible either we're doing something not quite right, or nobody's used this template in a long running function yet (which is quite possible since it's probably the fastest at the moment)

**Comment by brandonkal** [comment]
With this test function, I tested with a 20 sec value:
```
func Handle(req handler.Request) (handler.Response, error) {
	var err error

	time.Sleep(7 * time.Second)
	fmt.Printf("Current time %v", time.Now().Unix())
	message := fmt.Sprintf("Hello world, input was: %s", string(req.Body))

	return handler.Response{
		Body:       []byte(message),
		StatusCode: http.StatusOK,
	}, err
}
```

For my app, I am using 14400s.

**Comment by alexellis** [comment]
These values need to be made configurable in the specific template that you've chosen. a proposal for this on a specific repository would be welcome. 

What you might want to do in the interim is use the standard go template which already supports a configurable time out value.

**Comment by brandonkal** [comment]
@alexellis I have made a PR here:
https://github.com/openfaas-incubator/golang-http-template/pull/20
I did this all in the GitHub web ui os no signoff option. But I am mostly looking for feedback to see if the specific approach makes sense.

**Comment by alexellis** [comment]
Looks like this was already raised in October 2018, but nobody from the community volunteered to work on it.

https://github.com/openfaas-incubator/golang-http-template/issues/9

Alex

**Comment by brandonkal** [comment]
Unfortunately it appears this issue is still not resolved. I am running into a 10s timeout  even with my modifications to the golang-http-template.

**Comment by alexellis** [comment]
@brandonkal did you delete your comment? I just got an email from GitHub
